### PR TITLE
New math packages: cddlib, singular

### PR DESCRIPTION
**Pull request by dkwo**
New math packages, useful for sagemath

- New package: cddlib
- New package: singular

I built them on x86_64

### CODE REVIEW

**Comment by dkwo** [comment]
@tornaria could you take a look?
(it seems singular does not cross-compile, due to ntl.)

File: srcpkgs/singular/template [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# Template file for 'singular'
+pkgname=singular
+version=4.2.1
+revision=1
+_own_version=4-2-1
```
Comment (paper42) [inline srcpkgs/singular/template:5]: Could you use something like `${version//./-}`? Also remove this variable, because it's used only once.

**Comment by tornaria** [comment]
cddlib:
 - do we want to split a -devel pkg?
 - sage expects include files to be in /usr/include so this is not picked up. It seems old versions of cddlib used to place headers in usr/include, and debian places them in /usr/include/cdd, but recent versions of cddlib place them in /usr/include/cddlib.

**Comment by tornaria** [comment]
singular:
 - is it necessary to run autoreconf? it seems a configure is already provided in the source.

**Comment by tornaria** [comment]
More on singular:
 - can you enable readline for the command line?
 - we need to figure out if/how to split singular in pkgs: what's the minimum needed to run the command line?
 - it seems ntl is not being picked up; need to add ntl-devel to makedepends
 - python executable is searched in configure: I don't know what for.
 - dot seems to be needed to build graphics for the documentation (dot is in `graphviz`).

**Comment by dkwo** [comment]
>     * sage expects include files to be in /usr/include so this is not picked up. It seems old versions of cddlib used to place headers in usr/include, and debian places them in /usr/include/cdd, but recent versions of cddlib place them in /usr/include/cddlib.

This should be addressed in sage tickets 32243 and 29413 linked there, right?

**Comment by dkwo** [comment]
>     * python executable is searched in configure: I don't know what for.

Could this be due to `enable-python-module` option?

**Comment by dkwo** [comment]
@tornaria Thanks for the comments, I tried to implement some of them.
I've enabled readline to ncurses.

Do you undesrtand why `[target] ntl-devel-11.5.1_1: not found` for arm? is it a hostmakedep?

Can we postpone splitting to later?
I will perhaps try to address test failures first.

**Comment by dkwo** [comment]
Btw, the packages `ntl-devel` and `flintlib-devel` are not even found on the Void homepage - packages for `arm` and `aarch`, what am I missing here??

**Comment by tornaria** [comment]
> Btw, the packages `ntl-devel` and `flintlib-devel` are not even found on the Void homepage - packages for `arm` and `aarch`, what am I missing here? does it have to do with `nocross=yes` in ntl?

There is no native builder for arm, so `nocross` packages won't be available on arm.

Since `ntl` is optional to `flintlib` and singular, one option would be to disable ntl on cross builds. I was pointed to `vips` for an example: add a build_option `ntl` and have it enabled by default only if not cross-compiling. I'm not in a hurry to do that, let's get x86 working and later we can worry about cross.

**Comment by tornaria** [comment]
> > ```
> > * sage expects include files to be in /usr/include so this is not picked up. It seems old versions of cddlib used to place headers in usr/include, and debian places them in /usr/include/cdd, but recent versions of cddlib place them in /usr/include/cddlib.
> > ```
> 
> This should be addressed in sage tickets 32243 and 29413 linked there, right?

https://trac.sagemath.org/ticket/32243 seems to deal with this and already merged for sage-9.5 so let's just ignore this and wait (maybe it already works with sage-9.5.beta0, I didn't check).

**Comment by dkwo** [comment]
About musl test failures: on ci it fails `polys_test` for libpolys, while on my system (newer musl) it even fails `omtTest-m` and `omtTest-r` for omalloc (lots of `***omError_Unknown: unknown error`)

**Comment by dkwo** [comment]
> we need to figure out if/how to split singular in pkgs: what's the minimum needed to run the command line?

I think singualr comes command line only, right?

**Comment by tornaria** [comment]
> > we need to figure out if/how to split singular in pkgs: what's the minimum needed to run the command line?
> 
> I think singular comes command line only, right?

But it's also a library, and comes with headers, some static libraries, etc. Isn't that so? There's also lots of html documentation.

**Comment by dkwo** [comment]
> But it's also a library, and comes with headers, some static libraries, etc. Isn't that so? There's also lots of html documentation.

Sure, that we can definitely split.

I'm still trying to investigate the 2nd musl error, the one with polys.
Do you know whether it is possible to skip some tests and do others?
(I'm unable to read the log for polys, as my checks fail before, at omalloc.)

**Comment by dkwo** [comment]
Patch kindly provided by upstream.
Now we should decide how to split, and possibly which config options to use.
```
configure_args="
       --with-readline=ncurses
       --enable-bigintm-module
       --enable-bigintm-module
       --enable-Order-module
       --enable-gfanlib-module
       --disable-omalloc # this requires patch from upstream, useful for musl 1.2
       --enable-gfanlib
       --enable-Singular
       --enable-factory
       --disable-doc
       --disable-polymake
       --without-python
       --without-pythonmodule
       --disable-python
       --disable-python_module
       --disable-static
       --with-libparse"
```

**Comment by tornaria** [comment]
I Let's go with something standard that allows one to run `singular` and that we expect it will later work with sage. Is `--enable-bigintm-module` repeated above?

Regarding omalloc: I think it makes a difference in speed, so if not much trouble I would leave it enabled except when broken.

**Comment by dkwo** [comment]
Yes, that was a copy-paste of all possible options.
With current options (taken mostly from Sage), I can install and run `Singular`.
The omalloc test failure only occurs with newer musl (1.2), so I'll just leave a comment there,
as there's a chance it can get fixed upstream.

**Comment by dkwo** [comment]
@tornaria With the updated graphviz,  should the dependency here be modified?

**Comment by dkwo** [comment]
The omalloc test failure with newer musl has been fixed, so I think this PR is ready for review.
All patches are already merged upstream.
I did benefit a lot from the insights of @richfelker

**Comment by leahneukirchen** [comment]
For Singular, this is good enough to cross-build:

```
configure_args="--with-readline=ncurses
        --enable-gfanlib
        --enable-Singular
        --enable-factory
        --disable-doc
        --disable-polymake
        --without-python
        --with-libparse
        ac_cv_lib_cddgmp_dd_free_global_constants=yes
        "
```

For cddlib, please add cddlib-devel and also move the binaries that are not relevant for end-users there (or remove them completely).

**Comment by dkwo** [comment]
@leahneukirchen Done, thank you. Does it look better now?

**Comment by dkwo** [comment]
it seems cross-compilation now fails at build stage, with
```
In file included from gfanlib_zcone.cpp:19:
/usr/aarch64-linux-gnu/usr/include/cddlib/cdd.h:26:10: fatal error: cdd_f.h: No such file or directory
   26 | #include "cdd_f.h"
      |          ^~~~~~~~~
compilation terminated.
```
Am I still messing something up?

**Comment by tornaria** [comment]
For some reason the cross compiled version of cddlib is missing some files:
```
$ tar tvf hostdir/binpkgs/cdd/cddlib-devel-0.94m_1.aarch64.xbps 
-rw-r--r-- root/root       878 1969-12-31 21:00 ./props.plist
-rw-r--r-- root/root      2897 1969-12-31 21:00 ./files.plist
-rw-r--r-- 0/0             267 2021-11-09 07:39 ./usr/lib/pkgconfig/cddlib.pc
lrwxrwxrwx 0/0               0 2021-11-09 07:39 ./usr/lib/libcdd.so -> libcdd.so.0.1.2
-rw-r--r-- 0/0          241012 2021-11-09 07:39 ./usr/lib/libcdd.a
-rw-r--r-- 0/0             609 2021-11-09 07:39 ./usr/include/cddlib/splitmix64.h
-rw-r--r-- 0/0            1373 2021-11-09 07:39 ./usr/include/cddlib/setoper.h
-rw-r--r-- 0/0           10953 2021-11-09 07:39 ./usr/include/cddlib/cddtypes.h
-rw-r--r-- 0/0            4726 2021-11-09 07:39 ./usr/include/cddlib/cddmp.h
-rw-r--r-- 0/0           13933 2021-11-09 07:39 ./usr/include/cddlib/cdd.h
```
versus
```
$ tar tvf hostdir/binpkgs/cdd/cddlib-devel-0.94m_1.x86_64.xbps 
-rw-r--r-- root/root       877 1969-12-31 21:00 ./props.plist
-rw-r--r-- root/root      4267 1969-12-31 21:00 ./files.plist
-rw-r--r-- 0/0             276 2021-11-09 07:39 ./usr/lib/pkgconfig/cddlib.pc
lrwxrwxrwx 0/0               0 2021-11-09 07:39 ./usr/lib/libcddgmp.so -> libcddgmp.so.0.1.2
-rw-r--r-- 0/0          479286 2021-11-09 07:39 ./usr/lib/libcddgmp.a
lrwxrwxrwx 0/0               0 2021-11-09 07:39 ./usr/lib/libcdd.so -> libcdd.so.0.1.2
-rw-r--r-- 0/0          236468 2021-11-09 07:39 ./usr/lib/libcdd.a
-rw-r--r-- 0/0             609 2021-11-09 07:39 ./usr/include/cddlib/splitmix64.h
-rw-r--r-- 0/0            1373 2021-11-09 07:39 ./usr/include/cddlib/setoper.h
-rw-r--r-- 0/0           11307 2021-11-09 07:39 ./usr/include/cddlib/cddtypes_f.h
-rw-r--r-- 0/0           10953 2021-11-09 07:39 ./usr/include/cddlib/cddtypes.h
-rw-r--r-- 0/0            4947 2021-11-09 07:39 ./usr/include/cddlib/cddmp_f.h
-rw-r--r-- 0/0            4726 2021-11-09 07:39 ./usr/include/cddlib/cddmp.h
-rw-r--r-- 0/0           14714 2021-11-09 07:39 ./usr/include/cddlib/cdd_f.h
-rw-r--r-- 0/0           13933 2021-11-09 07:39 ./usr/include/cddlib/cdd.h
```

**Comment by tornaria** [comment]
I think it is because there is `hostmakedepends=gmp-devel` instead of `makedepends=gmp-devel`. Then configure is not finding gmp. Can you try again with that change?

**Comment by dkwo** [comment]
@tornaria Good point.
Now everything seems to run smoothly.

**Comment by tornaria** [comment]
LGTM: I compiled sage-9.5.beta5 which supports system singular out of the box.

**Comment by leahneukirchen** [comment]
@dkwo  Please remove the comments, the :shipit:

**Comment by dkwo** [comment]
@leahneukirchen Done!

**Pull request merged**